File: src/net/zomis/tttultimate/games/TTController.java
package net.zomis.tttultimate.games;

import net.zomis.tttultimate.OnMoveListener2;
import net.zomis.tttultimate.TTBase;
import net.zomis.tttultimate.TTPlayer;

public abstract class TTController {
	
	protected final TTBase game;
	protected TTPlayer currentPlayer = TTPlayer.X;
	private OnMoveListener2	moveListener;
	private StringBuilder history;

	public TTController(TTBase board) {
		this.game = board;
		this.history = new StringBuilder();
	}

	public abstract boolean isAllowedPlay(TTBase tile);
	
	public final boolean play(TTBase tile) {
		if (tile == null)
			throw new IllegalArgumentException("Tile to play at cannot be null.");
		
		if (!isAllowedPlay(tile)) {
			System.out.println("Warning: Move was not made. Unable to play at " + tile);
			return false;
		}
		TTBase playedTile = tile;
		if (!this.performPlay(tile))
			return false;

		this.addToHistory(tile);
		
		if (this.moveListener != null)
			this.moveListener.onMove(playedTile);
		
		return true;
	}
	
	private void addToHistory(TTBase tile) {
		if (history.length() > 0)
			history.append(",");
		history.append(Integer.toString(tile.getGlobalX(), Character.MAX_RADIX));
		history.append(Integer.toString(tile.getGlobalY(), Character.MAX_RADIX));
	}

	protected abstract boolean performPlay(TTBase tile);
	
	public boolean play(int x, int y) {
		return this.play(game.getSmallestTile(x, y));
	}
	
	public TTPlayer getCurrentPlayer() {
		return currentPlayer;
	}
	
	protected void nextPlayer() {
		currentPlayer = currentPlayer.next();
	}
	
	public TTBase getGame() {
		return game;
	}

	public boolean isGameOver() {
		return game.isWon();
	}

	public TTPlayer getWonBy() {
		return game.getWonBy();
	}

	public void setOnMoveListener(OnMoveListener2 moveListener) {
		this.moveListener = moveListener;
	}
	
	
	public void makeMoves(String history) throws IllegalStateException, IllegalArgumentException {
		for (String move : history.split(",")) {
			if (move.isEmpty())
				continue;
			if (move.length() != 2)
				throw new IllegalArgumentException("Unexcepted move length. " + move);
			
			int x = Integer.parseInt(String.valueOf(move.charAt(0)), Character.MAX_RADIX);
			int y = Integer.parseInt(String.valueOf(move.charAt(1)), Character.MAX_RADIX);
			
			TTBase tile = game.getSmallestTile(x, y);
			if (!this.play(tile))
				throw new IllegalStateException("Unable to make a move at " + x + ", " + y + ": " + tile);
		}
	}
	public String saveHistory() {
		return this.history.toString();
	}
	
	public void reset() {
		this.currentPlayer = TTPlayer.X;
		this.history = new StringBuilder();
		this.game.reset();
		this.onReset();
	}

	protected abstract void onReset();
	
	public String getViewFor(TTBase tile) {
		return tile.isWon() ? tile.getWonBy().toString() : "";
	}
}

File: uttt-common/src/main/kotlin/net/zomis/tttultimate/games/TTController.kt
package net.zomis.tttultimate.games

import net.zomis.tttultimate.TTBase
import net.zomis.tttultimate.TTMoveListener
import net.zomis.tttultimate.TTPlayer

abstract class TTController(val game: TTBase) {
    var currentPlayer = TTPlayer.X
        protected set
    private var moveListener: TTMoveListener? = null
    private var history: StringBuilder = StringBuilder()

    val CHARS = ('0'..'9') + ('a'..'z') + ('A'..'Z')

    val isGameOver: Boolean
        get() = game.isWon

    val wonBy: TTPlayer
        get() = game.wonBy

    abstract fun isAllowedPlay(tile: TTBase): Boolean

    fun play(tile: TTBase?): Boolean {
        if (tile == null)
            throw IllegalArgumentException("Tile to play at cannot be null.")

        if (!isAllowedPlay(tile)) {
            return false
        }
        if (!this.performPlay(tile))
            return false

        this.addToHistory(tile)

        if (this.moveListener != null)
            this.moveListener!!.onMove(tile)

        return true
    }

    private fun addToHistory(tile: TTBase) {
        if (!history.isEmpty()) {
            history.append(",")
        }
        history.append(CHARS[tile.globalX])
        history.append(CHARS[tile.globalY])
    }

    protected abstract fun performPlay(tile: TTBase): Boolean

    fun play(x: Int, y: Int): Boolean {
        return this.play(game.getSmallestTile(x, y))
    }

    protected fun nextPlayer() {
        currentPlayer = currentPlayer.next()
    }

    fun setOnMoveListener(moveListener: TTMoveListener) {
        this.moveListener = moveListener
    }

    fun makeMoves(history: String) {
        for (move in history.split(",")) {
            if (move.isEmpty())
                continue
            if (move.length != 2) {
                throw IllegalArgumentException("Unexcepted move length. $move")
            }

            val x = CHARS.indexOf(move[0])
            val y = CHARS.indexOf(move[1])

            val tile = game.getSmallestTile(x, y)
            if (!this.play(tile))
                throw IllegalStateException("Unable to make a move at $x, $y: $tile")
        }
    }

    fun saveHistory(): String {
        return this.history.toString()
    }

    fun reset() {
        this.currentPlayer = TTPlayer.X
        this.history = StringBuilder()
        this.game.reset()
        this.onReset()
    }

    protected abstract fun onReset()

    open fun getViewFor(tile: TTBase): String {
        return if (tile.isWon) tile.wonBy.toString() else ""
    }
}
